From 79fd44ad300166e30586c4e9b646db8d3a9f57bb Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Thu, 5 Dec 2024 14:36:53 +0800
Subject: [PATCH 179/274] AOSCOS: cpumask: resurrect cpumask_complement()

Torvalds himself removed this function in commit 596ff4a09b89 ("cpumask:
re-introduce constant-sized cpumask optimizations"), where he noted:

  In the process, remove 'cpumask_complement()' and 'for_each_cpu_not()'
  which were not useful, and which fundamentally have to be limited to
  'nr_cpu_ids'.  Better remove them now than have somebody introduce use
  of them later.

Well, apparently loongson64's disable_unused_cpus() found a use for this
function, so resurrect it until we find a better way to revise it.

Fixes: "AOSCOS: MIPS: Loongson 3: Introduce disable_unused_cpus() to save power"
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 include/linux/cpumask.h | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/include/linux/cpumask.h b/include/linux/cpumask.h
index 9278a50d514f..58eff4c9a6f0 100644
--- a/include/linux/cpumask.h
+++ b/include/linux/cpumask.h
@@ -682,6 +682,18 @@ bool cpumask_andnot(struct cpumask *dstp, const struct cpumask *src1p,
 					  cpumask_bits(src2p), small_cpumask_bits);
 }
 
+/**
+ * cpumask_complement - *dstp = ~*srcp
+ * @dstp: the cpumask result
+ * @srcp: the input to invert
+ */
+static inline void cpumask_complement(struct cpumask *dstp,
+				       const struct cpumask *srcp)
+{
+	bitmap_complement(cpumask_bits(dstp), cpumask_bits(srcp),
+					      nr_cpumask_bits);
+}
+
 /**
  * cpumask_equal - *src1p == *src2p
  * @src1p: the first input
-- 
2.47.1

